File: leetcode-python-go/117. 填充每个节点的下一个右侧节点指针 II/main.py
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

# 层序遍历解法
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        queue = [root]
        while queue:  # 遍历每一层
            length = len(queue)
            tail = None # 每一层维护一个尾节点
            for i in range(length): # 遍历当前层
                curnode = queue.pop(0)
                if tail:
                    tail.next = curnode # 让尾节点指向当前节点
                tail = curnode # 让当前节点成为尾节点
                if curnode.left : queue.append(curnode.left)
                if curnode.right: queue.append(curnode.right)
        return root

# 链表解法
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        first = root
        while first:  # 遍历每一层
            dummyHead = Node(None)  # 为下一行创建一个虚拟头节点，相当于下一行所有节点链表的头结点(每一层都会创建)；
            tail = dummyHead  # 为下一行维护一个尾节点指针（初始化是虚拟节点）
            cur = first
            while cur:  # 遍历当前层的节点
                if cur.left:  # 链接下一行的节点
                    tail.next = cur.left
                    tail = tail.next
                if cur.right:
                    tail.next = cur.right
                    tail = tail.next
                cur = cur.next  # cur同层移动到下一节点
            first = dummyHead.next  # 此处为换行操作，更新到下一行
        return root

# root = [1,2,3,4,5,"null",7]

sol = Solution()
root = Node(1, Node(2, Node(4), Node(5)), Node(3, None, Node(7)))
root = sol.connect(root)

result = []
node = root
while node:
    result.append(node.val)
    node = node.next
    if not node:
        result.append("#")
        node = root.left
        root = root.left

print(result)

File: leetcode-python-go/117. 填充每个节点的下一个右侧节点指针 II/main.go
package main

import (
	"container/list"
	"fmt"
	"strconv"
)

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	res := [][]*Node{}
	if root == nil { //防止为空
		return root
	}
	queue := list.New()
	queue.PushBack(root)
	var tmpArr []*Node
	for queue.Len() > 0 {
		length := queue.Len() //保存当前层的长度，然后处理当前层（十分重要，防止添加下层元素影响判断层中元素的个数）
		for i := 0; i < length; i++ {
			node := queue.Remove(queue.Front()).(*Node) //出队列
			if node.Left != nil {
				queue.PushBack(node.Left)
			}
			if node.Right != nil {
				queue.PushBack(node.Right)
			}
			tmpArr = append(tmpArr, node) //将值加入本层切片中
		}
		res = append(res, tmpArr) //放入结果集
		tmpArr = []*Node{}        //清空层的数据
	}
	//遍历每层元素,指定next
	for i := 0; i < len(res); i++ {
		for j := 0; j < len(res[i])-1; j++ {
			res[i][j].Next = res[i][j+1]
		}
	}
	return root
}

//添加遍历输出结果的代码
func traverse(root *Node) []string {
	res := []string{}
	queue := list.New()
	queue.PushBack(root)
	for queue.Len() > 0 {
		node := queue.Remove(queue.Front()).(*Node)
		if node.Left != nil {
			queue.PushBack(node.Left)
		}
		if node.Right != nil {
			queue.PushBack(node.Right)
		}
		if node.Next != nil {
			res = append(res, strconv.Itoa(node.Val), "#") //将整型转换成字符串
		} else {
			res = append(res, strconv.Itoa(node.Val))
		}
	}
	return res
}

func main() {
	root := &Node{Val: 1, Left: &Node{Val: 2, Left: &Node{Val: 4}, Right: &Node{Val: 5}}, Right: &Node{Val: 3, Left: nil, Right: &Node{Val: 7}}}
	connect(root)
	fmt.Println(traverse(root)) // 输出 [1, #, 2, 3, #, 4, 5, 7, #]
}
